src/components/tl/mlx5/mcast/tl_mlx5_mcast_progress.h

```diff
@@ -0,0 +1,89 @@
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast.h"
+#include "tl_mlx5_mcast_helper.h"
+
+#ifndef RELIABLE_H_
+#define RELIABLE_H_
+
+#define TO_VIRTUAL(_rank, _size, _root) ((_rank + _size - _root) % _size)
+
+#define TO_ORIGINAL(_rank, _size, _root) ((_rank + _root) % _size)
+
+#define ACK 1
+
+#define GET_COMPL_OBJ(_comm, _compl_fn, _pkt_id, _req)                          \
+    ({                                                                          \
+        void* item;                                                             \
+        ucc_tl_mlx5_mcast_p2p_completion_obj_t *obj;                            \
+        item = ucc_mpool_get(&(_comm)->ctx->compl_objects_mp);                  \
+        obj  = (ucc_tl_mlx5_mcast_p2p_completion_obj_t *)item;                  \
+                                                                                \
+        obj->data[0]  = (uintptr_t)_comm;                                       \
+        obj->compl_cb = _compl_fn;                                              \
+        obj->data[1]  = (uintptr_t)_pkt_id;                                     \
+        obj->data[2]  = (uintptr_t)_req;                                        \
+        obj;                                                                    \
+    })
+
+#define GET_NACK_REQ(_comm, _pkt_id)                                            \
+    ({                                                                          \
+        void* item;                                                             \
+        ucc_tl_mlx5_mcast_nack_req_t *_req;                                     \
+        item = ucc_mpool_get(&(_comm)->ctx->nack_reqs_mp);                      \
+                                                                                \
+        _req         = (ucc_tl_mlx5_mcast_nack_req_t *)item;                    \
+        _req->comm   = _comm;                                                   \
+        _req->pkt_id = _pkt_id;                                                 \
+        _req;                                                                   \
+    })
+
+#define P2P_OP(_comm, _buf, _size, _rank, _ctx, _op, _op_name, _compl_obj)      \
+    {                                                                           \
+        ucc_tl_mlx5_mcast_p2p_interface_t *p2p = &(_comm)->p2p;                 \
+        int ret =  _op ## _nb((_buf), (_size), (_rank),                         \
+                              (_ctx), _compl_obj);                              \
+        if (UCC_OK != ret) {                                                    \
+            tl_error(_comm->lib, "failed to do p2p "_op_name": comm %p,"        \
+                    " size %zu, rank %d",                                       \
+                      (_comm), (size_t)(_size), (_rank) );                      \
+            return UCC_ERR_NO_MESSAGE;                                          \
```

Comment:
Users of this P2P macro must be aware that this macro should be used in a function that returns a UCC status, otherwise, they cannot use it. In the future, if users need to handle errors of P2P in a different way, we can add a new macro or enhance what we have now